mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java

```diff
@@ -0,0 +1,129 @@
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.mr.hive.serde.objectinspector.IcebergWriteObjectInspector;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+class Deserializer {
+  private FieldDeserializer fieldDeserializer;
+
+  Deserializer(Schema schema, ObjectInspector fieldInspector) throws SerDeException {
+    this.fieldDeserializer = deserializer(schema.asStruct(), fieldInspector);
+  }
+
+  Record deserialize(Object data) {
+    return (Record) fieldDeserializer.value(data);
+  }
+
+  private interface FieldDeserializer {
+    Object value(Object object);
+  }
+
+  private static FieldDeserializer deserializer(Type type, ObjectInspector fieldInspector) throws SerDeException {
+    switch (type.typeId()) {
+      case BOOLEAN:
+        return o -> ((BooleanObjectInspector) fieldInspector).get(o);
+      case INTEGER:
+        return o -> ((IntObjectInspector) fieldInspector).get(o);
+      case LONG:
+        return o -> ((LongObjectInspector) fieldInspector).get(o);
+      case FLOAT:
+        return o -> ((FloatObjectInspector) fieldInspector).get(o);
+      case DOUBLE:
+        return o -> ((DoubleObjectInspector) fieldInspector).get(o);
+      case STRING:
+        return o -> ((StringObjectInspector) fieldInspector).getPrimitiveJavaObject(o);
+      case UUID:
+        // TODO: This will not work with Parquet. Parquet UUID expect byte[], others are expecting UUID
+        return o -> UUID.fromString(((StringObjectInspector) fieldInspector).getPrimitiveJavaObject(o));
+      case DATE:
+      case TIMESTAMP:
+      case FIXED:
+      case BINARY:
+      case DECIMAL:
+        // Iceberg specific conversions
+        return o -> ((IcebergWriteObjectInspector) fieldInspector).getIcebergObject(o);
+      case STRUCT:
+        return new StructDeserializer((Types.StructType) type, (StructObjectInspector) fieldInspector);
+      case LIST:
+      case MAP:
+      case TIME:
+      default:
+        throw new SerDeException("Unsupported column type: " + type);
+    }
+  }
+
+  private static class StructDeserializer implements FieldDeserializer {
+    private final FieldDeserializer[] fieldDeserializers;
+    private final StructObjectInspector fieldInspector;
+    private final Types.StructType type;
+
+    private StructDeserializer(Types.StructType type, StructObjectInspector fieldInspector) throws SerDeException {
+      List<? extends StructField> structFields = fieldInspector.getAllStructFieldRefs();
+      List<Types.NestedField> nestedFields = type.fields();
+      this.fieldDeserializers = new FieldDeserializer[structFields.size()];
+      this.fieldInspector = fieldInspector;
+      this.type = type;
+
+      for (int i = 0; i < fieldDeserializers.length; i++) {
+        fieldDeserializers[i] =
+            deserializer(nestedFields.get(i).type(), structFields.get(i).getFieldObjectInspector());
```

Comment:
As I noted above, I think this would be cleaner if you used a visitor and passed the field deserializers into this constructor instead of calling deserializer here.

Reply:
Done